### ISSUE TITLE
Flow: add Loki components

### ISSUE THREAD

**Issue by mattdurham** [grafana/agent#2499]
This issue proposes a set of `loki` components to add support for logs in Grafana Agent Flow pipelines. 
The new components should cover the same set of functionality that’s currently present in the ‘static’ agent mode when using the logs subsystem, by reusing Promtail code.

In spite of that, the proposal is to name these components after Loki because:
* The Loki name is more recognizable
* The protobuf format used for logs is itself from Loki, not Promtail
* It might be both a necessity and an opportunity to break some Promtail conventions when constructing pipelines, giving us more freedom.

## Proposal
We will create a new `loki` namespace to group the new logging components.
Additionally, a `loki.source` (sources?) namespace will be used to group the components which connect to log sources to pull log entries from, and a `loki.transform` namespace to group components that mutate the labels or the content of log entries.

## Pipeline construction
Flow pipelines work using expressions, where a component can reference the output of one another.

This worked well for both the metrics-specific Prometheus, as well as the otelcol.* set of components, so we can follow with the same approach.

The Promtail approach of using Go channels means that the various components will be synchronous; we can explore alternative approaches to fanout entries to multiple components that will not block the pipeline.

## Arguments
The subset of `loki` components which forward log entries to another component will support a `forward_to` argument which is a slice of the components where metrics should be fanned out to.

## Exported fields
The `loki` components which can receive log entries will support an exported field named `input` which will act similarly to how the `prometheus.Receiver` capsule value works today.

This deviates from the Prometheus components logic, but is more in line with how the OTel components work.

## Notes
* The Promtail architecture has not clearly defined boundaries between ‘service discovery’ and ‘scraping’. The target component implementations both discover applications who emit logs and send these entries over a Go channel.
* The Service Discovery mechanisms are only applicable for the FileTargetManager which only ever gets logs from the same host, so the main reason for running them is for getting additional labels and for filtering.
As such, we propose not having separate discovery components, nor reuse the components from the discovery namespace, but rather have the SD defined as blocks inside of loki.target.file
* Promtail uses the ‘pipelines’ terminology in a different way. To avoid confusion, we could focus on calling out the individual ‘stages’ that make them up.
* We should consider whether we want each component to keep a `position.yml` file in its own storage directory, allow the user to override it, or expliticly make them set up a global one.
* Currently in Promtail, when defining multiple Loki for remote writing, if the request to one Loki fails, it will impact the other Loki servers as well. Our approach to connecting components should avoid such behavior.

## Example configuration

```river
discovery.kubernetes "pods" { 
}

discovery.fileglob "pods" {
  targets      = discover.kubernetes.pods.targets
  path         = "/somepath"
  path_exclude = "/somepath/sensitive"
}

discovery.relabel {
  targets = discovery.fileglob.pods.targets
  rule { }
  rule { }
}

// Reads file from disk
loki.source.file "varlog" {
  targets      = concat(
	dicsovery.kubernetes.pods.output,
	{ "__path__" : "/var/log/nginx", "foo", "bar" },
  )
  
  forward_to =  [loki.transform.relabel.input]
}

// Runs relabeling rules to incoming log lines' labels.
loki.transform.relabel "default" {
  rule { /*  */ }
  rule { /*  */ }

  forward_to = [loki.transform.multiline.input]
}

// Runs Promtail stages to incoming log lines.
loki.transform "multiline" {
  stage { /*  */ }
  stage { /*  */ }

  forward_to = [loki.export.prod.input]
}

// Optional component that can act as an intermediary and flush data to disk as
// a pseudo-WAL until the WAL Promtail implementation is complete.
loki.wal "ondisk { }

// Send log lines over to a Loki instance.
loki.export "prod" {
  endpoint = “http://loki:3100”
  basic_auth {
     username = “foo”
     password = “bar”
  }
}
```

## Flow component list
Initially planned for v0.30.0:
- [x] #2503
- [x] #2514
- [x] #2515
- [x] #2516
- [x] #2518
- [x] #2525

For future release milestones:
- [x] #2505
- [x] #2506
- [x] #2507
- [x] #2508
- [x] #2509
- [x] #2510
- [x] #2511
- [x] #2512
- [x] #2692

**Comment by rfratto** [comment]
Do we need both otelcol.receiver.filelog and loki.target.file? Ditto with the two syslogs.

**Comment by tpaschalis** [comment]
> Do we need both otelcol.receiver.filelog and loki.target.file? Ditto with the two syslogs.

Not necessarily, and not the first priority (since logs already can come through the other OTel receivers), but I thought it'd be a good step in allowing people from both ecosystems to use pipelines without any transformations.

**Comment by rfratto** [comment]
> Not necessarily, and not the first priority (since logs already can come through the other OTel receivers), but I thought it'd be a good step in allowing people from both ecosystems to use pipelines without any transformations.

I'm personally leaning towards avoiding two implementations of something that does the same thing just to reduce the surface area where bugs may occur and reduce the amount of knowledge we have to track to offer support.

**Comment by tpaschalis** [comment]
> I'm personally leaning towards avoiding two implementations of something that does the same thing just to reduce the surface area where bugs may occur and reduce the amount of knowledge we have to track to offer support.

I've opened #2501 to discuss otelcol-specific logging components like these. I'll limit the initial scope of new components and we can revisit in the future if needed.

> I would vote for each component keeping track of its position in the storage directly allocated to it rather than letting users define their own.

We'll check how component storage works and when it's cleaned up. For example, if a component is temporarily removed due to invalid configuration and then comes back, does the positions file persist, or are the files tailed from the beginning?

> Should we use input more broadly?

I suppose you mean changing the metrics.Receiver name? I'd personally opt for _not_ changing something right away and reconsider once we got some logging components ready.

> I'd like to avoid using the remote write name to avoid being confused with the remote_write protocol. Can this be named something else, like export or client?

I see your point. In this case, I think the `export` name goes well in hand with people familiar with the ETL.

**Comment by rfratto** [comment]
> For example, if a component is temporarily removed due to invalid configuration and then comes back, does the positions file persist, or are the files tailed from the beginning?

At the moment, the data directory created for a component never gets deleted. 

> I suppose you mean changing the metrics.Receiver name? I'd personally opt for not changing something right away and reconsider once we got some logging components ready.

👍 SGTM

**Comment by rfratto** [comment]
> ```river
> // Creates and manages a Promtail FileTargetManager to read files from disk.
> loki.target.file "varlog" {
>  path         = "/somepath/"
>  path_exclude = "/somepath/sensitive"
>  filter       = loki.target.filter.staging.output
>  
>  kubernetes_sd  { /*  */ } // Adds labels from the Kubernetes API server to log lines.
>  
>  forward_to =  [loki.transform.relabel.input]
> }
> ```

I think we should feel comfortable forking Promtail code to give us more control over the behavior of log collection. Given that, we're not beholden to how FileTargetManager works today, and that feels like a huge opportunity to redesign how logs from files get connected.

Is it possible for us to do a more Prometheus-esque approach for collecting logs from files? (i.e., SD targets -> log "scrape" job)

The one complexity this might cause is how Promtail supports expending file globs to multiple targets, which is probably why FileTargetManager works the way it does.

**Comment by mattdurham** [comment]
IMO before making big changes I am going to start on a component this afternoon/morning and get a feel for how it flows. I also like the prometheus method but I think the scrape gets more complex because it is not an HTTP endpoint but could be a huge number of formats/sources.

**Comment by rfratto** [comment]
> Is it possible for us to do a more Prometheus-esque approach for collecting logs from files? (i.e., SD targets -> log "scrape" job)
>
> The one complexity this might cause is how Promtail supports expending file globs to multiple targets, which is probably why FileTargetManager works the way it does.

(Thought off the top of my head: one way to do this is to have a discovery component which can expand file path globs, though that would usually mean it always needs to sit in front of file tailing)

**Comment by tpaschalis** [comment]
After some discussion, there have been some changes to the proposal. I will update the issue titles and descriptions so that they are self-descriptive, but to summarize:

* We opt to use the 'sources' and 'sinks' terminology instead of 'targets' and 'remote_write'.
* We will try to be more generic in our SD components approach. The 'sources' components will expect a set of targets, and will not do service discovery themselves (eg. expanding fileglobs and watching directories).
* We will introduce a discovery.fileglob component which will expand glob patterns and enrich them with a list of labels
* We've separated the list of proposed components into ones that we'll try to tackle in the upcoming v0.30 release, and ones that can wait for the next one.

**Comment by tpaschalis** [comment]
Update: The v0.30.0 release candidate is out with the initial planned set of components as planned.

The next steps for the v0.31.0 release are
a) to polish the current set of components with smarter handling of targets 
b) to iterate on `loki.process` and build the complete set of functionality around processing log entries,
c) to work on #2501 to implement converters between OTel and Loki,
d) to start implementing more log sources in order of popularity/usefulness.

**Comment by rfratto** [comment]
All of the initial set of Loki components have been implemented, where the last of them will be released in v0.32. 

Nice work, everyone!
